cli/tools/script/shell_parser.rs

```diff
@@ -0,0 +1,769 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::anyhow::bail;
+use deno_core::error::AnyError;
+
+use crate::tools::script::combinators::ParseErrorFailure;
+
+use super::combinators::assert;
+use super::combinators::assert_exists;
+use super::combinators::char;
+use super::combinators::delimited;
+use super::combinators::if_not;
+use super::combinators::many0;
+use super::combinators::many_till;
+use super::combinators::map;
+use super::combinators::maybe;
+use super::combinators::or;
+use super::combinators::separated_list;
+use super::combinators::skip_whitespace;
+use super::combinators::tag;
+use super::combinators::take_while;
+use super::combinators::terminated;
+use super::combinators::with_error_context;
+use super::combinators::ParseError;
+use super::combinators::ParseResult;
+
+// Shell grammar rules this is loosely based on:
+// https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html#tag_02_10_02
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct SequentialList {
+  pub items: Vec<SequentialItem>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum SequentialItem {
+  Command(ShellCommand),
+  Async(ShellCommand),
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum ListOperator {
+  // &&
+  And,
+  // ||
+  Or,
+}
+
+impl ListOperator {
+  pub fn as_str(&self) -> &'static str {
+    match self {
+      ListOperator::And => "&&",
+      ListOperator::Or => "||",
+    }
+  }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct ShellCommand {
+  pub body: ShellCommandBody,
+  pub next: Option<Box<NextCommand>>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum ShellCommandBody {
+  /// Command like `MY_VAR=5`
+  EnvVar(EnvVar),
+  Command(Command),
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Command {
+  pub env_vars: Vec<EnvVar>,
+  pub args: Vec<String>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct NextCommand {
+  pub op: ListOperator,
+  pub command: ShellCommand,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct EnvVar {
+  pub name: String,
+  pub value: String,
+}
+
+impl EnvVar {
+  pub fn new(name: String, value: String) -> Self {
+    EnvVar { name, value }
+  }
+}
+
+/// Note: Only used to detect redirects in order to give a better error.
+/// Redirects are not part of the first pass of this feature.
+pub struct Redirect {
+  pub maybe_fd: Option<usize>,
+  pub op: RedirectOp,
+  pub word: Option<String>,
+}
+
+pub enum RedirectOp {
+  /// >
+  Redirect,
+  /// >>
+  Append,
+}
+
+pub fn parse(input: &str) -> Result<SequentialList, AnyError> {
+  fn error_for_failure(
+    e: ParseErrorFailure,
+  ) -> Result<SequentialList, AnyError> {
+    bail!(
+      "{}\n  {}\n  ~",
+      e.message,
+      e.input.chars().take(60).collect::<String>()
```

Comment:
Why 60?

Reply:
I can add a comment. This is just a simple implementation that truncates to prevent word wrap.